NiL.JS/Core/Tools.cs

```diff
@@ -1887,5 +1887,393 @@ internal static LambdaExpression BuildJsCallTree(string name, Expression functio
+
+                        s = new StringBuilder($"Array ({len}) [ ");
+                        int i = 0;
+                        for (i = 0; i < len; i++)
+                        {
+                            if (i > 0)
+                                s.Append(", ");
+                            s.Append(Tools.JSValueToObjectString(a[i], maxRecursionDepth, recursionDepth + 1));
+                        }
+                        if (a._fields != null)
+                        {
+                            for (var e = a._fields.GetEnumerator(); e.MoveNext();)
+                            {
+                                if (i++ > 0)
+                                    s.Append(", ");
+                                s.Append(e.Current.Key).Append(": ");
+                                s.Append(Tools.JSValueToObjectString(e.Current.Value, maxRecursionDepth, recursionDepth + 1));
+                            }
+                        }
+                        s.Append(" ]");
+
+                        return s.ToString();
+                    }
+                    else
+                    {
+                        string typeName = Tools.GetTypeName(v);
+                        if (recursionDepth >= maxRecursionDepth)
+                            return typeName;
+
+                        StringBuilder s = new StringBuilder(typeName);
+                        s.Append(" { ");
+
+                        JSObject o = v as JSObject;
+                        if (o == null)
+                            o = v._oValue as JSObject;
+                        if (o == null)
+                            return v.ToString();
+
+
+                        int i = 0;
+                        for (var e = o.GetEnumerator(true, EnumerationMode.RequireValues); e.MoveNext();)
+                        {
+                            if (i++ > 0)
+                                s.Append(", ");
+                            s.Append(e.Current.Key).Append(": ");
+                            s.Append(Tools.JSValueToObjectString(e.Current.Value, maxRecursionDepth, recursionDepth + 1));
+                        }
+
+                        s.Append(" }");
+
+                        return s.ToString();
+                    }
+            }
+            return v.ToString();
+        }
+        
+        public static string JSValueToString(JSValue v)
+        {
+            if (v == null)
+                return "null";
+
+            if (v.ValueType == JSValueType.Object)
+            {
+                if (v._oValue == null)
+                    return "null";
+
+                var o = v as ObjectWrapper;
+                if (o == null)
+                    o = v._oValue as ObjectWrapper;
+                if (o != null)
+                    return o.Value.ToString();
+            }
+
+            return v.ToString();
+        }
+
+
+        public static string FormatArgs(IEnumerable<KeyValuePair<string, JSValue>> args)
```

Comment:
On word on new public functions:
All except for one of the functions I added to Tools are public; Do you want to make those internal as well? Some of them could be useful, e.g. in my little console, I use JSValueToObjectString with a high maxRecursionDepth to display output.
Also, functions like GetTypeName could be useful but are currently a little improvised. It works great for what it is used now but we might be able to break it.
Speaking of GetTypeName, maybe we should replace the code in JSValue.toString for objects with it to have the code in a central place.